File: Python/programmers/etc/level3/야근_지수.py
from heapq import *

def solution(n, works):
    q = []
    for w in works:
        heappush(q, -w)
    while n:
        if q[0] >= 0:
            break
        w = heappop(q)
        heappush(q, w+1)
        n -= 1
        
    return sum([i**2 for i in q])


works = [4, 3, 3]
n = 4
print(solution(works, n)) # 12

works = [2, 1, 2]
n = 1
print(solution(works, n)) # 6

works = [1, 1]
n = 3
print(solution(works, n)) # 0

File: Java/programmers/etc/level3/야근지수.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 야근지수 {
    public long solution(int n, int[] works) {
        long answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for (int work : works) {
            pq.add(-work);
        }
        
        while (n != 0) {
            if (pq.peek() >= 0) break;
            int a = pq.poll();
            pq.add(a+1);
            n--;
        }
        
        for (Integer a : pq) {
            answer += a*a;
        }
        
        return answer;
    }
}
